accepted/2022/simplify-output-paths.md

```diff
@@ -0,0 +1,95 @@
+# Simplify .NET Output Paths
+
+We'd like to simplify the output paths for .NET 8.
+
+Currently, the default output path for a .NET project includes folders for the `TargetFramework` and the `RuntimeIdentifier` (if it is set).  Additionally, the publish output goes in a `publish` subfolder of the output folder.  The resulting paths are as follows:
+
+- `bin\<Configuration>\<TargetFramework>` - Build output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>` - Build output with `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\publish` - Publish output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>\publish` - Publish output with `RuntimeIdentifier`
+
+This is rather complicated and inconsistent, which can make for a poor first impression of the .NET platform.
+
+We'd like to try to improve the output path structure.  Desired qualities include:
+
+- Simple and consistent
+- Avoid cases where one output path is nested inside of another one
+- Avoid excessively deep directory hierarchies
+- Get rid of `obj` folder in project root
+
+## Proposed behavior
+
+The new output path format will consist of the following 3 nested folders:
+
+- `artifacts` - All output will go under this folder
+- Output Type - Such as `bin`, `publish`, `intermediates`, or `package`
+- Pivots - This will at minimum be the `Configuration` (lower-cased), such as `debug` or `release`. Other pivots such as `TargetFramework` or `RuntimeIdentifier` may also be included, and the pivots will be joined by the underscore (`_`) character
+  - `TargetFramework` will be included in the folder name if the project is multi-targeted (`TargetFrameworks` is non-empty), or if the `TargetFramework` property was set via the command line (ie is a global property)
+  - `RuntimeIdentifier` will be included in the folder name if it was explicitly set (either in a project file or on the command line).  If it is set automatically by the SDK (for example because `SelfContained` was set), then the `RuntimeIdentifier` will not be included in the folder name
+  - For the `package` output type, the pivots will only be the `Configuration`, and other values won't be included in the path
+
+Some examples:
+
+- `artifacts\bin\debug` - The build output path for a simple project when you run `dotnet build`
```

Comment:
What happens when a project starts as targeting a single TFM, performs a build, and then is modified to be multi-targeted? Will the relevant artifacts directory be cleared before performing any build to avoid it containing both layouts in that scenario?

Reply:
Good point. I suspect that no one wants or will appreciate multiple layouts in a single build.

Reply:
I agree that no one wants their build output directories to be a mess. In general, incremental builds are not resilient against substantial configuration changes. It is a best practice to run `git clean -xdf` after any substantial configuration changes.

Reply:
I think the key issue with that though is it's not universally understood (or even intuitive) which changes constitute "substantial configuration changes".

Reply:
Yep, it leads to a behavior where people run `git clean` every time there is remote chance of stale content in output directories causing problems. It is cheaper and predictable to learn to run `git clean` than spend hours debugging the problem with stale content in your output directory every once in a while. For example, the suggested daily workflow for dotnet/runtime starts with `git clean`: https://github.com/dotnet/runtime/blob/main/docs/workflow/building/libraries/README.md#quick-start

Reply:
While I appreciate your point, I don't think we should settle on that being the "norm" for .NET as a development experience. We should be extremely careful when using the runtime repo as any kind of benchmark for experience.

Reply:
> Yep, it leads to a behavior where people run git clean every time there is remote chance of stale content in output directories causing problems.

True for developers like us who work on .NET itself and have to deal with complex build graphs but probably less of a concern for the average developers with small to medium sized solutions. IMO I don't see any strong reason why a non-multi-targeting build's outputs must be layered differently than a multi-targeting build's outputs.

I don't think that it's uncommon to change between `TargetFramework` and `TargetFrameworks` and I would expect the .NET SDK handle that for me without "poisoning" my build.

Reply:
We should also avoid taking a dependency on source control functionality for basic operations. It makes sense for the dotnet/runtime use case, but shouldn't be what we rely on for anything with this proposal.

Reply:
FWIW, I have seen many issues over the years where people poisoned their output directory while doing simple things. Probably the most common issue is passing different extra properties via the command line for different build invocations. It is possible that my view is biased due to only seeing the cases that are broken.

Reply:
I haven gotten myself into busted states that I didn't understand. I usually do a bunch of `rm -rf bin obj`, which fixes the problem. Definitely sympathetic.